server/routes/user_router.go

```diff
@@ -0,0 +1,358 @@
+package routes
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/caitlin615/nist-password-validator/password"
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/rawdaGastan/grid3_auto_deployer/internal"
+	"github.com/rawdaGastan/grid3_auto_deployer/models"
+)
+
+type SignUpInput struct {
+	Name            string `json:"name" binding:"required"`
+	Email           string `json:"email" gorm:"unique" binding:"required"`
+	Password        string `json:"password" binding:"required"`
+	ConfirmPassword string `json:"confirmPassword" binding:"required"`
+}
+
+type AuthDataInput struct {
+	Email string `json:"email"`
+	Code  int    `json:"code"`
+}
+
+type ChangePasswordInput struct {
+	Email           string `json:"email"`
+	Password        string `json:"password" binding:"required"`
+	ConfirmPassword string `json:"confirmPassword" binding:"required"`
+}
+
+func (router *Router) SignUpHandler(w http.ResponseWriter, r *http.Request) {
+	var signUp SignUpInput
+	err := json.NewDecoder(r.Body).Decode(&signUp)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	// validate mail
+	bool := internal.ValidateMail(signUp.Email)
+	if !bool {
+		router.WriteErrResponse(w, fmt.Errorf("email isn't valid %v", err))
+		return
+	}
+
+	u := models.User{
+		Name:     signUp.Name,
+		Email:    signUp.Email,
+		Password: signUp.Password,
+	}
+
+	// check if user already exists
+	_, err = router.db.GetUserByEmail(u.Email, router.secret)
+	if err == nil {
+		router.WriteMsgResponse(w, fmt.Errorf("user already exists"))
+	}
+
+	// password should be ACII , min 5 , max 10
+	validator := password.NewValidator(true, 5, 10)
+	err = validator.ValidatePassword(signUp.Password)
+	if err != nil {
+		router.WriteErrResponse(w, fmt.Errorf("error: %v password isn't valid", err))
+		return
+	}
+
+	// password and confirm password should match
+	if signUp.Password != signUp.ConfirmPassword {
+		router.WriteErrResponse(w, fmt.Errorf("password and confirm password don't match"))
+		return
+	}
+
+	// hash password
+	hashedPassword, err := internal.HashPassword(u.Password)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+	u.Password = hashedPassword
+
+	// send verification code
+	code, err := internal.SendMail(u.Email)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+	u.Code = code
+	fmt.Printf("code: %v\n", code)
+	msg := "Verfification Code has been sent to " + u.Email
+	router.db.SetCache(u.Email, u)
```

Comment:
why saving all user data. I think we need code and email only

Reply:
how to get all data entered by user to create the account ? 
That's why i saved it all